cli/internal/helm/backup_test.go

```diff
@@ -0,0 +1,140 @@
+			require := require.New(t)
+			memFs := afero.NewMemMapFs()
+
+			crd := apiextensionsv1.CustomResourceDefinition{}
+			err := yaml.Unmarshal([]byte(tc.crd), &crd)
+			require.NoError(err)
+			client := Client{config: nil, kubectl: stubCrdClient{crds: []apiextensionsv1.CustomResourceDefinition{crd}, getCRDsError: tc.getCRDsError}, fs: memFs, log: stubLog{}}
+
+			_, err = client.backupCRDs(context.Background())
+			if tc.wantError {
+				assert.Error(err)
+				return
+			}
+			assert.NoError(err)
+
+			data, err := afero.ReadFile(memFs, filepath.Join(crdBackupFolder, crd.Name+".yaml"))
+			require.NoError(err)
+			assert.YAMLEq(tc.expectedFile, string(data))
+		})
+	}
+}
+
+func TestBackupCRs(t *testing.T) {
+	testCases := map[string]struct {
+		crd          apiextensionsv1.CustomResourceDefinition
+		resource     unstructured.Unstructured
+		expectedFile string
+		getCRsError  error
+		wantError    bool
+	}{
+		"success": {
+			crd:          apiextensionsv1.CustomResourceDefinition{Spec: apiextensionsv1.CustomResourceDefinitionSpec{Names: apiextensionsv1.CustomResourceDefinitionNames{Plural: "foobars"}, Group: "some.group", Versions: []apiextensionsv1.CustomResourceDefinitionVersion{{Name: "versionZero"}}}},
+			resource:     unstructured.Unstructured{Object: map[string]any{"metadata": map[string]any{"name": "foobar"}}},
+			expectedFile: "metadata:\n  name: foobar\n",
+		},
+		"api request fails": {
+			crd:         apiextensionsv1.CustomResourceDefinition{Spec: apiextensionsv1.CustomResourceDefinitionSpec{Names: apiextensionsv1.CustomResourceDefinitionNames{Plural: "foobars"}, Group: "some.group", Versions: []apiextensionsv1.CustomResourceDefinitionVersion{{Name: "versionZero"}}}},
+			getCRsError: errors.New("api error"),
+			wantError:   true,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			assert := assert.New(t)
+			require := require.New(t)
+			memFs := afero.NewMemMapFs()
+
+			client := Client{config: nil, kubectl: stubCrdClient{crs: []unstructured.Unstructured{tc.resource}, getCRsError: tc.getCRsError}, fs: memFs, log: stubLog{}}
+
+			err := client.backupCRs(context.Background(), []apiextensionsv1.CustomResourceDefinition{tc.crd})
+			if tc.wantError {
+				assert.Error(err)
+				return
+			}
+			assert.NoError(err)
+
+			data, err := afero.ReadFile(memFs, filepath.Join(backupFolder, tc.resource.GetName()+".yaml"))
+			require.NoError(err)
+			assert.YAMLEq(tc.expectedFile, string(data))
+		})
+	}
+}
+
+type stubLog struct{}
+
+func (s stubLog) Debugf(format string, args ...any) {}
+func (s stubLog) Sync()                             {}
+
+type stubCrdClient struct {
+	crds         []apiextensionsv1.CustomResourceDefinition
+	getCRDsError error
+	crs          []unstructured.Unstructured
+	getCRsError  error
+}
+
+func (c stubCrdClient) Initialize(kubeconfig []byte) error {
+	return nil
+}
+
+func (c stubCrdClient) ApplyCRD(ctx context.Context, rawCRD []byte) error {
+	return nil
+}
```

Comment:
```suggestion
type stubCrdClient struct {
	crds         []apiextensionsv1.CustomResourceDefinition
	getCRDsError error
	crs          []unstructured.Unstructured
	getCRsError  error
	crdClient
}
```
You can embed the `crdClient` interface to save on defining unused functions for the testing struct.
Note that this will panic if one of the functions is called (which might be useful when the code is updated in the future to make use of them).